### PR TITLE
Project-specific package curations

### PR DESCRIPTION
**Pull request by MarcelBochtler**
Enable package curations from the project's repository configuration file in order to enable project-specific curations.

### CODE REVIEW

File: model/src/main/kotlin/config/Curations.kt [hunk 1]
```diff
@@ -21,10 +21,18 @@ package org.ossreviewtoolkit.model.config
...
+
 /**
  * Curations for artifacts in a repository.
  */
 data class Curations(
+    /**
+     * Curations for third-party packages.
+     */
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    val packageCurations: List<PackageCuration> = emptyList(),
```
Comment (fviernau) [inline model/src/main/kotlin/config/Curations.kt:34]: For consistency I believe it should be named `packages` which is a reserved word.
How about naming it `packagez` or `pkgs` and override the serialized name.

File: docs/config-file-ort-yml.md [hunk 2]
```diff
@@ -4,7 +4,8 @@ The items below can be configured by adding an `.ort.yml` file to the root of th
 
 * [excludes](#excludes) - Mark [files, directories](#excluding-paths) or [package manager scopes](#excluding-scopes) as
   not included in released artifacts.
-* [license finding curations](#curations) - Overwrite scan results to correct identified licenses.
+* [curations](#curations) - Overwrite scan results to correct identified licenses or correct metadata of third-party
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:7]: How about stealing the first sentence from `config-file-curations-yml.md´? e.g. "Curations correct invalid or missing package metadata and set the concluded license for packages."

File: docs/config-file-ort-yml.md [hunk 3]
```diff
@@ -125,11 +126,10 @@ scopes defined in the examples below match the scopes in your project.
 
 ### When to Use Curations
 
-Project-specific curations should be used when you want to correct the licenses detected in the source code of the
-project. If you need to correct the license findings for a third-party dependency then add a curation to
-[curations.yml](config-file-curations-yml.md) or [package configuration](config-file-package-configuration-yml.md).
+Project-specific license finding curations should be used when you want to correct the licenses detected in the source
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:129]: 'license finding curations' != curations

I believe this is out of scope of this PR.

File: docs/config-file-ort-yml.md [hunk 4]
```diff
@@ -125,11 +126,10 @@ scopes defined in the examples below match the scopes in your project.
...
 ### When to Use Curations
 
-Project-specific curations should be used when you want to correct the licenses detected in the source code of the
-project. If you need to correct the license findings for a third-party dependency then add a curation to
-[curations.yml](config-file-curations-yml.md) or [package configuration](config-file-package-configuration-yml.md).
+Project-specific license finding curations should be used when you want to correct the licenses detected in the source 
+code of the project.
 
-### Curating License Findings
+### Curating Project License Findings
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:132]: Why add `Project`? License finding curations apply also to `Packages`.

File: docs/config-file-ort-yml.md [hunk 5]
```diff
@@ -156,6 +156,25 @@ For details of the specification, see
 The list of available options for `reason` are defined in
 [LicenseFindingCurationReason.kt](../model/src/main/kotlin/config/LicenseFindingCurationReason.kt).
 
+### Curating Third-Party Metadata
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:159]: Curations are not limited to third-Party packages but could also curated your orgs' packages.

File: docs/config-file-ort-yml.md [hunk 6]
```diff
@@ -156,6 +156,25 @@ For details of the specification, see
...
+
+Project-specific package curations can be added if you want to correct metadata of third-party dependencies.
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:161]: Instead of introducing multiple sections speaking about "project specific" things, how about explaining that once in the first section of this file. E.g. that all configuration in this file applies only to this Project's context. Usually global is preferred for increased degree of automation and local should only be done if there is a good reason.

**Review by fviernau — CHANGES_REQUESTED** [review]

**Comment by fviernau** [comment]
Note: the PR title says "package configurations" but it actually means "package curations"

File: model/src/main/kotlin/config/Curations.kt (hunk 1 above)
Comment (MarcelBochtler) [reply model/src/main/kotlin/config/Curations.kt:34]: Only `package` is a reserved word. The plural is not. I'll change it.

File: docs/config-file-ort-yml.md (hunk 3 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:129]: > 'license finding curations' != curations

I am aware.

The removed sentence: `If you need to correct the license findings for a third-party dependency then add a curation to
[curations.yml](config-file-curations-yml.md) or [package configuration](config-file-package-configuration-yml.md).`

Is now no longer true. As third-party dependencies can now also be curated in the `.ort.yml`.
I think this should be part of this PR.

File: docs/config-file-ort-yml.md (hunk 4 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:132]: The documentation states:
> Project-specific curations should be used when you want to correct the licenses detected in the source code of the
project.

Which implicitly excludes packages.
The [used Model](https://github.com/oss-review-toolkit/ort/blob/e9204dc54768b07967873d7ecd3506d5a4c1a1dc/model/src/main/kotlin/config/LicenseFindingCuration.kt#L33-L78) doesn't define an `Identifier`.
How can the described `license_finding_curations` in this case apply to packages?

File: docs/config-file-ort-yml.md (hunk 5 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:159]: Removed `Third-Party`.

File: docs/config-file-ort-yml.md (hunk 6 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:161]: Done in a separate commit

File: docs/config-file-ort-yml.md (hunk 3 above)
Comment (fviernau) [reply docs/config-file-ort-yml.md]: I meant "Project-specific license finding curations" don't exist, not even in this PR.

File: docs/config-file-ort-yml.md (hunk 4 above)
Comment (fviernau) [reply docs/config-file-ort-yml.md:133]: You're right, my bad.

File: docs/config-file-ort-yml.md (hunk 6 above)
Comment (fviernau) [reply docs/config-file-ort-yml.md]: Thx!

File: docs/config-file-ort-yml.md [hunk 7]
```diff
@@ -1,6 +1,8 @@
 # The `.ort.yml` file
 
 The items below can be configured by adding an `.ort.yml` file to the root of the source code repository.
+All configurations in this file apply only to this Project's context. Usually the global context is preferred for an
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:4]: Commit-msg: "...is used **for**..."

File: docs/config-file-ort-yml.md [hunk 8]
```diff
@@ -6,7 +6,8 @@ increased degree of automation and local configurations should only be done if t
 
 * [excludes](#excludes) - Mark [files, directories](#excluding-paths) or [package manager scopes](#excluding-scopes) as
   not included in released artifacts.
-* [license finding curations](#curations) - Overwrite scan results to correct identified licenses.
+* [curations](#curations) - Curations correct invalid or missing package metadata and set the concluded license for
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:9]: The structure is changed now from

```
curations:
  license_findings:
```
to
```
curations:
  license_findings:
  packages:
```
The description now only describes package curations. I believe it should describe both.
Maybe with "Overwrite package meta-data, set a concluded license or correct license findings." or maybe separate sections for `package curations` and `license finding curations` would be more straight forward?

File: docs/config-file-ort-yml.md [hunk 9]
```diff
@@ -127,11 +128,9 @@ scopes defined in the examples below match the scopes in your project.
 
 ### When to Use Curations
 
-Curations should be used when you want to correct the licenses detected in the source code of the
-project. If you need to correct the license findings for a third-party dependency then add a curation to
-[curations.yml](config-file-curations-yml.md) or [package configuration](config-file-package-configuration-yml.md).
+License finding curations should be used when you want to correct the licenses detected in the source code of the project.
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:131]: This section is about both, package curations and license finding curations. Is it? ..but it only mentions the latter.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: docs/config-file-ort-yml.md (hunk 9 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:131]: Yes. I'll add the part about package configurations.

File: docs/config-file-ort-yml.md (hunk 3 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md]: > I meant "Project-specific license finding curations" don't exist, not even in this PR.

Project-specific license finding curations do exist as per the documentation: https://github.com/oss-review-toolkit/ort/blob/master/docs/config-file-ort-yml.md#curating-license-findings
Or at least a functionality is called like that in the documentation.

Are you talking about project-specific package configurations with a license finding curation?
Yes. This is out of the scope for this PR.

Comment (fviernau) [reply docs/config-file-ort-yml.md]: > Are you talking about project-specific package configurations with a license finding curation?
Yes. This is out of the scope for this PR.

no, I'm talking about the the elements under
``
curations:
  license_findings:
``

Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md]: This was previously and still is documented:
`License finding curations should be used when you want to correct the licenses detected in the source code of the
project.`

Should I rephrase this to clarify this?

File: docs/config-file-ort-yml.md [hunk 10]
```diff
@@ -127,11 +127,11 @@ scopes defined in the examples below match the scopes in your project.
 
 ### When to Use Curations
 
-Curations should be used when you want to correct the licenses detected in the source code of the
-project. If you need to correct the license findings for a third-party dependency then add a curation to
-[curations.yml](config-file-curations-yml.md) or [package configuration](config-file-package-configuration-yml.md).
+License finding curations should be used when you want to correct the licenses detected in the source code of the
+project. To define path excludes on a global level for third-party packages, please use the
+[package configurations](config-file-package-configuration-yml.md).
```
Comment (fviernau) [inline docs/config-file-ort-yml.md:132]: Why is the second sentence explaining  path excludes? e.g. the previous section is about excludes.
I'm not saying it's wrong, just thinking out loud: Should this section only be above the types of curations which can be put into the ort.yml as this file is about the ort.yml?

If yes, then this section should explain (1) license finding curations (2) package curations.

I'd also be fine to leave it as is in this PR, but make a follow up PR to extract the "When to Use Curations" out of this file and move it to a more general place.  The concept of "package curations", "license finding curations" and "path excludes" is  or gets now implemented both 'project-local' and 'global'. The answer to the when to use question is same no matter if "local" or "global" so it shouldn't be in this documentation about "local" configuration.

What do you think?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: docs/config-file-ort-yml.md (hunk 9 above)
Comment (fviernau) [reply docs/config-file-ort-yml.md]: I believe you mixed up "package curations" with "package configurations".

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 11]
```diff
@@ -170,18 +171,21 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
...
+        val repositoryConfiguration = actualRepositoryConfigurationFile?.readValueOrNull() ?: RepositoryConfiguration()
+
         val curationProvider = FallbackPackageCurationProvider(
             listOfNotNull(
                 FilePackageCurationProvider.from(packageCurationsFile, packageCurationsDir),
                 config.analyzer.sw360Configuration?.let {
                     Sw360PackageCurationProvider(it).takeIf { useSw360Curations }
                 },
+                SimplePackageCurationProvider(repositoryConfiguration.curations.packages).takeIf {
+                    repositoryConfiguration.curations.packages.isNotEmpty()
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:183]: This fallback provider is taking only the curations from the first provider in the list which has at least one.

I believe that the first item in the list should be a provider which return the union of local and global package curations.
Both, global and local configuration should always apply. BTW I believe that having mulitple curations per package is perfectly fine, we should not assume that the global configuration only contains one curation per package, nor should we assume the same for the local configuration and also not for their union.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: docs/config-file-ort-yml.md (hunk 10 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:132]: > I'd also be fine to merge this PR as-is, but make a follow up PR to extract the "When to Use Curations" out of this file and move it to a more general place.

As this documentation will be rewritten again as soon as _project-local_ package configurations are made possible, I'd prefer to do this in a separate PR.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt (hunk 11 above)
Comment (MarcelBochtler) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt]: Done.
Please also see the added description in the commit message.

Regarding the order. What should be the curation that takes precedence if it's conflicting with another one?
With the implemented change it is local > global because this is how it was discussed for a similar discussion with `license choices` with @mnonnenmacher.

Comment (fviernau) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt]: I agree with the order.

Comment (fviernau) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt]: Thanks!

**Review by fviernau — APPROVED** [review]
